skfda/misc/metrics/_sdtw_distances.py

```diff
@@ -0,0 +1,346 @@
+
+from skfda._utils._utils import _check_compatible_fdata
+from skfda.misc.metrics import _sdtw_numba
+from skfda.misc._math import inner_product_matrix, inner_product
+
+
+def _check_shape_postive_cost_mat(
+    cost: NDArrayFloat,
+    expected_shape: Tuple,
+    shape_only: Boolean = False
+) -> None:
+    """check that matrix 'cost' is nonnegative and has expected shape"""
+    n1, n2 = expected_shape
+    # must have shape fdata1.grid_points[0], fdata1.grid_points[0]
+    if not cost.shape == (n1, n2):
+        raise ValueError(
+            f"Cost matrix must have shape ({n1}, {n2})"
+        )
+    # non-negativity
+    if not(shape_only):
+        if not np.all(cost >= 0):
+            raise ValueError(
+                "Cost matrix must contain non-negative entries"
+            )
+
+
+def _check_indiscernable_sym(
+    cost: NDArrayFloat,
+    X: NDArrayFloat,
+    Y: NDArrayFloat
+) -> None:
+    """check that cost returns a symmetric matrix with zero diag"""
+    # indiscernability: cost(x_t, x_t) = 0 for any t
+    if (
+        np.all(np.diag(cost(X, X)) != 0)
+        or np.all(np.diag(cost(Y, Y)) != 0)
+    ):
+        raise ValueError(
+            "The cost between two identical objects "
+            "must be zero (i.e cost(X, X) is zero diagonal)"
+        )
+
+    # symmetry: cost(X, Y) = cost(y, X).T
+    if not np.allclose(cost(X, Y), cost(Y, X).T):
+        raise ValueError(
+            "The cost function does not return a symmetric matrix"
+        )
+
+
+def _check_input_fdata(
+    fdata1,
+    fdata2
+) -> None:
+    """check compatible, one sample and one-dimensional"""
+    _check_compatible_fdata(fdata1, fdata2)
+
+    if not isinstance(fdata1, FDataGrid):
+        raise ValueError(
+            "fdata1 and fdata2 must be FDataGrid objects"
+        )
+
+    # dim_codomain can be higher than one,
+    # but dim_domain must be one
+    if not fdata1.dim_domain == 1:
+        raise ValueError(
+            "fdata1 and fdata2 must have their domain dimension equal to one"
+        )
+
+    if not (fdata1.n_samples == 1 & fdata2.n_samples == 1):
+        raise ValueError(
+            "Both fdata1 and fdata2 must each contain a single sample"
+        )
+
+
+def half_sq_euclidean(
+    arg1: NDArrayFloat,
+    arg2: Optional[NDArrayFloat] = None,
+) -> NDArrayFloat:
+    """
+    Return the half squared Euclidean row distance between
+    two matrices of size (n1, d) and (n2, d).
+
+    Args:
+        arg1: First sample (n_samples_1, n_features).
+        arg2: Second sample (n_samples_2, n_features).
+            If it is ``None`` returns the half squared Euclidean distance
+            between the rows of ``arg1``.
+
+    The distance is taken w.r.t matrix rows.
+    """
+    if arg2 is not None:
+        # np.dot(X, Y.T)=inner_product_matrix(X, Y) but np ~1.6 faster
+        # X.shape is (n_X, d) and Y.shape is (n_Y, d)
+        cost_12 = -1 * inner_product_matrix(arg1, arg2)
```

Comment:
Why catastrophic ?

Reply:
Okay!  I haven't got it sorry

Reply:
The problem is that with the factorized version 0.5 * inner_product_matrix(X-Y, X-Y) instead, we cannot compute the cost between two FDataGrid  that have different grid size, which is breaks a major advantage of SDTW

Reply:
Or maybe you were asking for a safer way of coding the product ?!

Reply:
I think that currently `inner_product` and `inner_product_matrix` are only implemented for `FDataGrid`s with the same grid points. Are you proposing to implement that for different grids (but same domain range)? It could be done relying on the same numerical integration as general functions, such as `FDataBasis`. It would probably be less exact and slower in that case, though.

If we are ok with that and with catastrophic cancellation, then I agree that it is better this way.

Reply:
Yes it can deal with different grids, assuming same domain range for ``FDaGrid``s. Okay so I let it as it is.